[Compiled Autograd] Turn accumulate_grad into an op

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111274
* #111273
* __->__ #111271


Rather than baking the behavior of `AccumulateGrad` nodes into the generated graph (either as `+=`, or as a return value of the graph).  This creates a new `accumulate_grad_` dispatcher op that is included in the generated graph like:
```
def forward(self, inputs, sizes, hooks):
    getitem = inputs[0]
    getitem_1 = inputs[1]
    getitem_2 = inputs[2]
    getitem_3 = inputs[3]
    getitem_4 = inputs[4]
    getitem_5 = inputs[5]
    getitem_6 = inputs[6]
    getitem_7 = inputs[7]
    getitem_8 = inputs[8]
    getitem_9 = inputs[9];  inputs = None
    expand = torch.ops.aten.expand.default(getitem, [2, 4]);  getitem = None
    threshold_backward = torch.ops.aten.threshold_backward.default(expand, getitem_1, 0);  expand = getitem_1 = None
    t = torch.ops.aten.t.default(getitem_3);  getitem_3 = None
    mm = torch.ops.aten.mm.default(threshold_backward, t);  t = None
    t_1 = torch.ops.aten.t.default(threshold_backward)
    mm_1 = torch.ops.aten.mm.default(t_1, getitem_2);  t_1 = getitem_2 = None
    t_2 = torch.ops.aten.t.default(mm_1);  mm_1 = None
    sum_1 = torch.ops.aten.sum.dim_IntList(threshold_backward, [0], True);  threshold_backward = None
    view = torch.ops.aten.view.default(sum_1, [4]);  sum_1 = None
    t_3 = torch.ops.aten.t.default(t_2);  t_2 = None
    accumulate_grad_ = torch.ops.inductor.accumulate_grad_.default(getitem_4, t_3);  getitem_4 = t_3 = None
    threshold_backward_1 = torch.ops.aten.threshold_backward.default(mm, getitem_5, 0);  mm = getitem_5 = None
    t_4 = torch.ops.aten.t.default(threshold_backward_1)
    mm_2 = torch.ops.aten.mm.default(t_4, getitem_6);  t_4 = getitem_6 = None
    t_5 = torch.ops.aten.t.default(mm_2);  mm_2 = None
    sum_2 = torch.ops.aten.sum.dim_IntList(threshold_backward_1, [0], True);  threshold_backward_1 = None
    view_1 = torch.ops.aten.view.default(sum_2, [4]);  sum_2 = None
    t_6 = torch.ops.aten.t.default(t_5);  t_5 = None
    accumulate_grad__1 = torch.ops.inductor.accumulate_grad_.default(getitem_7, t_6);  getitem_7 = t_6 = None
    accumulate_grad__2 = torch.ops.inductor.accumulate_grad_.default(getitem_8, view_1);  getitem_8 = view_1 = None
    accumulate_grad__3 = torch.ops.inductor.accumulate_grad_.default(getitem_9, view);  getitem_9 = view = None
    return []

```

The motivation here is `AccumulateGrad` nodes are causing trouble in FSDP tracing, since FSDP is in-place resizing parameters and parameter storage in hooks.  We will model this mutation in dynamo, but not during the initial compiled autograd capture.  This allows us to bypass failing shape checks in the initial capture. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

jansel committed:
[Compiled Autograd] Turn accumulate_grad into an op

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Turn accumulate_grad into an op"



Rather than baking the behavior of `AccumulateGrad` nodes into the generated graph (either as `+=`, or as a return value of the graph).  This creates a new `accumulate_grad_` dispatcher op that is included in the generated graph like:
```
def forward(self, inputs, sizes, hooks):
    getitem = inputs[0]
    getitem_1 = inputs[1]
    getitem_2 = inputs[2]
    getitem_3 = inputs[3]
    getitem_4 = inputs[4]
    getitem_5 = inputs[5]
    getitem_6 = inputs[6]
    getitem_7 = inputs[7]
    getitem_8 = inputs[8]
    getitem_9 = inputs[9];  inputs = None
    expand = torch.ops.aten.expand.default(getitem, [2, 4]);  getitem = None
    threshold_backward = torch.ops.aten.threshold_backward.default(expand, getitem_1, 0);  expand = getitem_1 = None
    t = torch.ops.aten.t.default(getitem_3);  getitem_3 = None
    mm = torch.ops.aten.mm.default(threshold_backward, t);  t = None
    t_1 = torch.ops.aten.t.default(threshold_backward)
    mm_1 = torch.ops.aten.mm.default(t_1, getitem_2);  t_1 = getitem_2 = None
    t_2 = torch.ops.aten.t.default(mm_1);  mm_1 = None
    sum_1 = torch.ops.aten.sum.dim_IntList(threshold_backward, [0], True);  threshold_backward = None
    view = torch.ops.aten.view.default(sum_1, [4]);  sum_1 = None
    t_3 = torch.ops.aten.t.default(t_2);  t_2 = None
    accumulate_grad_ = torch.ops.inductor.accumulate_grad_.default(getitem_4, t_3);  getitem_4 = t_3 = None
    threshold_backward_1 = torch.ops.aten.threshold_backward.default(mm, getitem_5, 0);  mm = getitem_5 = None
    t_4 = torch.ops.aten.t.default(threshold_backward_1)
    mm_2 = torch.ops.aten.mm.default(t_4, getitem_6);  t_4 = getitem_6 = None
    t_5 = torch.ops.aten.t.default(mm_2);  mm_2 = None
    sum_2 = torch.ops.aten.sum.dim_IntList(threshold_backward_1, [0], True);  threshold_backward_1 = None
    view_1 = torch.ops.aten.view.default(sum_2, [4]);  sum_2 = None
    t_6 = torch.ops.aten.t.default(t_5);  t_5 = None
    accumulate_grad__1 = torch.ops.inductor.accumulate_grad_.default(getitem_7, t_6);  getitem_7 = t_6 = None
    accumulate_grad__2 = torch.ops.inductor.accumulate_grad_.default(getitem_8, view_1);  getitem_8 = view_1 = None
    accumulate_grad__3 = torch.ops.inductor.accumulate_grad_.default(getitem_9, view);  getitem_9 = view = None
    return []

```

The motivation here is `AccumulateGrad` nodes are causing trouble in FSDP tracing, since FSDP is in-place resizing parameters and parameter storage in hooks.  We will model this mutation in dynamo, but not during the initial compiled autograd capture.  This allows us to bypass failing shape checks in the initial capture. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel committed:
Update on "[Compiled Autograd] Turn accumulate_grad into an op"



Rather than baking the behavior of `AccumulateGrad` nodes into the generated graph (either as `+=`, or as a return value of the graph).  This creates a new `accumulate_grad_` dispatcher op that is included in the generated graph like:
```
def forward(self, inputs, sizes, hooks):
    getitem = inputs[0]
    getitem_1 = inputs[1]
    getitem_2 = inputs[2]
    getitem_3 = inputs[3]
    getitem_4 = inputs[4]
    getitem_5 = inputs[5]
    getitem_6 = inputs[6]
    getitem_7 = inputs[7]
    getitem_8 = inputs[8]
    getitem_9 = inputs[9];  inputs = None
    expand = torch.ops.aten.expand.default(getitem, [2, 4]);  getitem = None
    threshold_backward = torch.ops.aten.threshold_backward.default(expand, getitem_1, 0);  expand = getitem_1 = None
    t = torch.ops.aten.t.default(getitem_3);  getitem_3 = None
    mm = torch.ops.aten.mm.default(threshold_backward, t);  t = None
    t_1 = torch.ops.aten.t.default(threshold_backward)
    mm_1 = torch.ops.aten.mm.default(t_1, getitem_2);  t_1 = getitem_2 = None
    t_2 = torch.ops.aten.t.default(mm_1);  mm_1 = None
    sum_1 = torch.ops.aten.sum.dim_IntList(threshold_backward, [0], True);  threshold_backward = None
    view = torch.ops.aten.view.default(sum_1, [4]);  sum_1 = None
    t_3 = torch.ops.aten.t.default(t_2);  t_2 = None
    accumulate_grad_ = torch.ops.inductor.accumulate_grad_.default(getitem_4, t_3);  getitem_4 = t_3 = None
    threshold_backward_1 = torch.ops.aten.threshold_backward.default(mm, getitem_5, 0);  mm = getitem_5 = None
    t_4 = torch.ops.aten.t.default(threshold_backward_1)
    mm_2 = torch.ops.aten.mm.default(t_4, getitem_6);  t_4 = getitem_6 = None
    t_5 = torch.ops.aten.t.default(mm_2);  mm_2 = None
    sum_2 = torch.ops.aten.sum.dim_IntList(threshold_backward_1, [0], True);  threshold_backward_1 = None
    view_1 = torch.ops.aten.view.default(sum_2, [4]);  sum_2 = None
    t_6 = torch.ops.aten.t.default(t_5);  t_5 = None
    accumulate_grad__1 = torch.ops.inductor.accumulate_grad_.default(getitem_7, t_6);  getitem_7 = t_6 = None
    accumulate_grad__2 = torch.ops.inductor.accumulate_grad_.default(getitem_8, view_1);  getitem_8 = view_1 = None
    accumulate_grad__3 = torch.ops.inductor.accumulate_grad_.default(getitem_9, view);  getitem_9 = view = None
    return []

```

The motivation here is `AccumulateGrad` nodes are causing trouble in FSDP tracing, since FSDP is in-place resizing parameters and parameter storage in hooks.  We will model this mutation in dynamo, but not during the initial compiled autograd capture.  This allows us to bypass failing shape checks in the initial capture. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

voznesenskym commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
```diff
@@ -83,14 +84,11 @@ variable_list AccumulateGrad::apply_with_saved(
-        saved.assign_mutable_grad(variable, grad_update);
-      });
+  variable_copy.mutable_grad() = grad_copy;
+  static auto op = c10::Dispatcher::singleton()
```
nit: If this is intentionally static for lifetime reasons, comment. I often find stuff like this gets lost in future refactors and programs subtly change.

jansel replied:
Isn't this the standard pattern for using this dispatcher?  Or is there another way to do it?

voznesenskym replied:
not sure, static seems safe here, but still worth a comment.

jansel replied:
I added a comment

voznesenskym commented on torch/csrc/inductor/inductor_ops.cpp:
```diff
@@ -38,13 +39,33 @@ Tensor _reinterpret_tensor(
+        grad,
+        new_grad,
+        1 /* num_expected_refs */,
+        [&grad](at::Tensor&& grad_update) { grad = std::move(grad_update); });
```
If we have a lifecycle preserving analogue of std::reference_wrapper that raises on access, that would be nice, because then if this lambda is stored or used after grad goes out of scope, we don't get undefined behavior.

That being said, I'm kinda assuming that this is just invoked immediately based on how you wrote it?

jansel replied:
This is the exact same thing accumulate_grad nodes do:
https://github.com/pytorch/pytorch/blob/d7317d8a1172dce9871803ff4b2d92d1acf869d6/torch/csrc/autograd/functions/accumulate_grad.cpp#L58

jansel committed:
Update on "[Compiled Autograd] Turn accumulate_grad into an op"



Rather than baking the behavior of `AccumulateGrad` nodes into the generated graph (either as `+=`, or as a return value of the graph).  This creates a new `accumulate_grad_` dispatcher op that is included in the generated graph like:
```
def forward(self, inputs, sizes, hooks):
    getitem = inputs[0]
    getitem_1 = inputs[1]
    getitem_2 = inputs[2]
    getitem_3 = inputs[3]
    getitem_4 = inputs[4]
    getitem_5 = inputs[5]
    getitem_6 = inputs[6]
    getitem_7 = inputs[7]
    getitem_8 = inputs[8]
    getitem_9 = inputs[9];  inputs = None
    expand = torch.ops.aten.expand.default(getitem, [2, 4]);  getitem = None
    threshold_backward = torch.ops.aten.threshold_backward.default(expand, getitem_1, 0);  expand = getitem_1 = None
    t = torch.ops.aten.t.default(getitem_3);  getitem_3 = None
    mm = torch.ops.aten.mm.default(threshold_backward, t);  t = None
    t_1 = torch.ops.aten.t.default(threshold_backward)
    mm_1 = torch.ops.aten.mm.default(t_1, getitem_2);  t_1 = getitem_2 = None
    t_2 = torch.ops.aten.t.default(mm_1);  mm_1 = None
    sum_1 = torch.ops.aten.sum.dim_IntList(threshold_backward, [0], True);  threshold_backward = None
    view = torch.ops.aten.view.default(sum_1, [4]);  sum_1 = None
    t_3 = torch.ops.aten.t.default(t_2);  t_2 = None
    accumulate_grad_ = torch.ops.inductor.accumulate_grad_.default(getitem_4, t_3);  getitem_4 = t_3 = None
    threshold_backward_1 = torch.ops.aten.threshold_backward.default(mm, getitem_5, 0);  mm = getitem_5 = None
    t_4 = torch.ops.aten.t.default(threshold_backward_1)
    mm_2 = torch.ops.aten.mm.default(t_4, getitem_6);  t_4 = getitem_6 = None
    t_5 = torch.ops.aten.t.default(mm_2);  mm_2 = None
    sum_2 = torch.ops.aten.sum.dim_IntList(threshold_backward_1, [0], True);  threshold_backward_1 = None
    view_1 = torch.ops.aten.view.default(sum_2, [4]);  sum_2 = None
    t_6 = torch.ops.aten.t.default(t_5);  t_5 = None
    accumulate_grad__1 = torch.ops.inductor.accumulate_grad_.default(getitem_7, t_6);  getitem_7 = t_6 = None
    accumulate_grad__2 = torch.ops.inductor.accumulate_grad_.default(getitem_8, view_1);  getitem_8 = view_1 = None
    accumulate_grad__3 = torch.ops.inductor.accumulate_grad_.default(getitem_9, view);  getitem_9 = view = None
    return []

```

The motivation here is `AccumulateGrad` nodes are causing trouble in FSDP tracing, since FSDP is in-place resizing parameters and parameter storage in hooks.  We will model this mutation in dynamo, but not during the initial compiled autograd capture.  This allows us to bypass failing shape checks in the initial capture. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

voznesenskym commented:
```
[2023-10-17T14:28:50.156-07:00] Buck UI: https://www.internalfb.com/buck2/a4a98015-edc7-493d-91f4-4c3fba8a97df
[2023-10-17T14:28:57.911-07:00] Waiting on buck2 daemon a4a98015-edc7-493d-91f4-4c3fba8a97df...
[2023-10-17T14:28:57.911-07:00] Resource usage: RSS: 5.5GiB CPU: 4492%
[2023-10-17T14:28:57.911-07:00] IO: none
[2023-10-17T14:29:05.012-07:00] Waiting on fbcode//sigrid/trainer/ads/data_pipeline/stages:filters (ovr_config//platform/linux:x86_64-fbcode-platform010-clang15-split-dwarf-CUDA-projects-cuda-12-cuda-12-python-3.8#a7d161811995e56a) -- running analysis [evaluate_rule], and 41 other actions
[2023-10-17T14:29:06.081-07:00] RE Session: reSessionID-8c2a7115-71d2-441c-aa94-8b3f537d39b7
[2023-10-17T14:29:13.134-07:00] Waiting on fbcode//dper3/dper3_models/experimental/pytorch/simple_model/simple_sparsenn_model:simple_sparsenn_scripting_test (ovr_config//platform/linux:x86_64-fbcode-platform010-clang15-split-dwarf-CUDA-projects-cuda-12-cuda-12-python-3.8#a7d161811995e56a) -- running analysis [evaluate_rule], and 678 other actions
[2023-10-17T14:29:20.162-07:00] Waiting on fbcode//dper3/dper3_models/experimental/pytorch/tests:ads_model_generator_utils_test (ovr_config//platform/linux:x86_64-fbcode-platform010-clang15-split-dwarf-CUDA-projects-cuda-12-cuda-12-python-3.8#a7d161811995e56a) -- running analysis [evaluate_rule], and 2033 other actions
[2023-10-17T14:29:27.191-07:00] Waiting on fbcode//dper3/dper3_models/experimental/pytorch/ppml_si/split_net:parameter_aggregation_split_test (ovr_config//platform/linux:x86_64-fbcode-platform010-clang15-split-dwarf-CUDA-projects-cuda-12-cuda-12-python-3.8#a7d161811995e56a) -- running analysis [evaluate_rule], and 6423 other actions
[2023-10-17T14:29:34.219-07:00] Waiting on fbcode//dper3/dper3_models/experimental/pytorch/tests:plot_dper_model_test (ovr_config//platform/linux:x86_64-fbcode-platform010-clang15-split-dwarf-CUDA-projects-cuda-12-cuda-12-python-3.8#a7d161811995e56a) -- running analysis [evaluate_rule], and 3350 other actions
[2023-10-17T14:29:41.279-07:00] Waiting on fbcode//pytorch/captum:install (ovr_config//platform/linux:x86_64-fbcode-platform010-clang15-split-dwarf-CUDA-projects-cuda-12-cuda-12-python-3.8#a7d161811995e56a) -- action (genrule) [re_download], and 4967 other actions
[2023-10-17T14:29:45.356-07:00] Action failed: fbcode//caffe2:aot-inductor-c-libtorch (ovr_config//platform/linux:x86_64-fbcode-platform010-clang15-split-dwarf-CUDA-projects-cuda-12-cuda-12-python-3.8#a7d161811995e56a) (cxx_compile torch/csrc/inductor/inductor_ops.cpp (pic))
[2023-10-17T14:29:45.356-07:00] Remote command returned non-zero exit code 1
[2023-10-17T14:29:45.356-07:00] Remote action, reproduce with: `frecli cas download-action fd28bdd7948d7271edf7c1c123f76713a3a9f5b8ebf4a77ba7ebd311927beafa:145`
[2023-10-17T14:29:45.356-07:00] Stdout: <empty>
[2023-10-17T14:29:45.356-07:00] Stderr:
fbcode/caffe2/torch/csrc/inductor/inductor_ops.cpp:7:10: fatal error: 'torch/csrc/autograd/functions/accumulate_grad.h' file not found
#include <torch/csrc/autograd/functions/accumulate_grad.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

```
Weird build failure https://www.internalfb.com/sandcastle/workflow/3742491290346126838/artifact/actionlog.3742491290371233288.stdout.1?selectedLines=133-153-1-1 

godspeed @jansel lmk if you need a hand
